### ISSUE TITLE
How should we provide visual feedback for Vimium's mode?

### ISSUE THREAD

**Issue by smblott-github** [philc/vimium#1484]
In 1.49, in insert mode and find mode visual feedback is provided by the presence (or absence) of the HUD.

In `master`, we currently use badges, like this...

![snapshot](https://cloud.githubusercontent.com/assets/2641335/6202566/c4a3b804-b4ea-11e4-9e75-f80ffe9f3b9b.png)

There is considerable discussion in #1413 as to:
- Is this the best approach?
- What if the user has the Vimium icon hidden?

[cVim](https://github.com/1995eaton/chromium-vim) uses a pop-up, like this:

![snapshot](https://cloud.githubusercontent.com/assets/2641335/6202580/5e2fedbc-b4eb-11e4-83ae-41a38aed7ea1.png)

So:  What is the best approach?

Me?  I prefer badges, perhaps unsurprisingly:
- It's the chrome way.
- It feels more integrated with chrome.
- It doesn't take up any extra screen real estate.
- Since there's no badge in normal mode, the presence or absence of a badge is noticeable out of the corner of the eye.
- Unlike the HUD, it doesn't appear in odd places.
- We look up at the page icon anyway to see if passkeys are active.
- The HUD takes up a lot of space for about two or three bits of information.
- We could use the [chrome notification API](https://developer.chrome.com/apps/notifications) to inform users that there's significance to the Vimium page icon.
- Other popular extensions provide their visual feedback in a similar way (e.g. AdBlock Plus, various mail checkers, etc.)

So:  What should we do?

**Comment by z0rch** [comment]
I like the badges too. 

I would only ask for the entire Vimium icon to change its color depending on the mode, so that a user could tell the current mode by a quick glance, without focusing on that little letter in the badge.

**Comment by mrmr1993** [comment]
I would like to go back to the HUD. I'm not a fan of the badges, and feel strongly that it was a bad decision for UI/UX. My criticisms are (sorry for the essay this became):

#### TL;DR:

* We're assuming we're more important than we are for a lot of people, and cutting a feature for those that disagree (because obviously they're misguided).
* We're using badges in a way that they're not intended (they're supposed to be annoying, so you do whatever it is to get rid of them).
* This is a bad way to communicate the information.

-----
#### Original post

* We force the user to show the browser icon:
  - we permanently use a block of screen estate, but are usually showing no useful information.
  - we've moved from a low value area of the screen to stealing space from the omnibox, which is much higher value. This is especially painful on narrow windows.
  - there is no alternative for users without the icon.
  - fullscreen windows have no information about the mode.
* Badges are small, and the abbreviations are cryptic, so more physical/mental effort.
  - At a glance, it's only easy to tell that the extension has *something* to tell us, not what.
  - We lose the benefit of automatic recognition of words by shape, and meaning is harder to extract.
* The badge is in a different 'context' than the information it's describing:
  - the badge is unrelated to the function of the button.
  - we're making the user pay attention to the top bar when it's not otherwise relevant, making them widen their field of focus across (extra) UI boundaries.
* Always-on badges have a reputation (imho rightly) for being irritating, and are used for this reason to pester users to do something.
  - In use, I can feel this effect: I actively want to get back into normal mode to get rid of the badge.
  - The iOS OTA upgrade badge on settings was a good, widespread example of this.
* We have no explaination anywhere for what the badge is/what the codes mean.
  - We're aspiring to create an intuitice user experience. Having to work out what it is, and then learn and memorise these codes on the go is counter to this.
* We intend to educate our users that we're important enough to feature constantly in a prominant position. We're not that important to everyone.

> * It's the chrome way.

I've never seen chrome use/recommend a badge to indicate the mode of something, only that there is an action available to the user. Can you provide an example of this?

> * It feels more integrated with chrome.

Chrome's philosophy is to get out of the way until needed; to be available but not intrusive. I certainly wouldn't say that Chrome's in-page shortcuts (scrolling, tabbing between inputs, accesskeys, selection commands, etc.) feel 'Chromey', and nor should they.

> * It doesn't take up any extra screen real estate.

I disagree. It *permanently* repurposes screen real estate from something more important (omnibox).

> Other popular extensions provide their visual feedback in a similar way

Again, I'm not familiar with any popular extensions which describe their current mode using an icon badge. A few use the icon/a badge to indicate that they are enabled/disabled, but that seems to be the extent of it. Also, no other popular extensions seem to be under the impression that all users must and will have their icon permanently visible.

**Comment by z0rch** [comment]
> (badges) are supposed to be annoying, so you do whatever it is to get rid of them

This is not always like that. To give you an example, LastPass (a password manager) shows a badge if there is a password stored for the current website. You are not supposed to get rid of this badge, it's just a piece of useful information, that is shown on the icon. 

Personally I find this indicator useful so many times, that I prefer LastPass icon to be always visible. I see no problem for myself to make Vimium icon permanently visible if it provides such useful information as well.

> > It doesn't take up any extra screen real estate.
>
> I disagree. It permanently repurposes screen real estate from something more important (omnibox).

I don't think we will be able to agree on which real estate is more valuable, icons row or website. My vote goes for website, even though I agree that on small screens having many icons permanently visible is complicated. 

Much more important is that this information should stand out better, which you also mention:

> Badges are small, and the abbreviations are cryptic, so more physical/mental effort.

This is my personal leader of annoyance as of today. I like the powerline/airline plugins for vim, so naturally I keep suggesting to incorporate the use of colors to display the current mode, like these:

![powerline](https://camo.githubusercontent.com/3c3a1717e42f17651f688ecc19f87e7433275098/68747470733a2f2f7261772e6769746875622e636f6d2f706f7765726c696e652f706f7765726c696e652f646576656c6f702f646f63732f736f757263652f5f7374617469632f696d672f706c2d6d6f64652d6e6f726d616c2e706e67)
![powerline](https://camo.githubusercontent.com/627cc0d6ca618d9480dfbeeaf1f35ca0cc30781b/68747470733a2f2f7261772e6769746875622e636f6d2f706f7765726c696e652f706f7765726c696e652f646576656c6f702f646f63732f736f757263652f5f7374617469632f696d672f706c2d6d6f64652d696e736572742e706e67)
![powerline](https://camo.githubusercontent.com/4a4454311274d16299aef08227496c5d1a7a9fd5/68747470733a2f2f7261772e6769746875622e636f6d2f706f7765726c696e652f706f7765726c696e652f646576656c6f702f646f63732f736f757263652f5f7374617469632f696d672f706c2d6d6f64652d76697375616c2e706e67)
![powerline](https://camo.githubusercontent.com/28430e34155892705de259c0e5fb0eec63825856/68747470733a2f2f7261772e6769746875622e636f6d2f706f7765726c696e652f706f7765726c696e652f646576656c6f702f646f63732f736f757263652f5f7374617469632f696d672f706c2d6d6f64652d7265706c6163652e706e67)

> We force the user to show the browser icon:
> 
> * we permanently use a block of screen estate, but are usually showing no useful information.

Personally I find useful to understand which mode I am in, so even though 90% of time the icon would say "normal mode", it is still useful for me.

**Comment by saheel1115** [comment]
How about implementing the HUD and giving the user the option to hide/disable it? And you can keep the badges too for those who like icons. Yes, I understand that might take up more memory.

Not sure how much a user's views matter, but here's 2 cents anyway:

1. Screen real estate is really precious for me on my 13 inch screen. Thus, many a times I prefer to use Full Screen, which hides the icon and leaves me with no information on Vimium mode. Also, I disagree with the point of real estate being made against HUD. I feel a small occasional HUD pop-up does not take up so much real estate. This is very subjective, though. Nonetheless, from an utilitarian POV, space freed by Full Screen - space taken by HUD >> 0.

2. I prefer to remove badges and keep my address bar clean, otherwise there'll be a plethora of them. Currently, I have badges for AdBlock and Vimium because I don't have a choice. If you guys decide to go the HUD way, I'll surely hide the Vimium icon.

3. This is subjective but reading the badge is definitely tougher for me, colored or otherwise. Being an Emacs user (and this must be true for Vim users as well), I am just so used to reading the status bar that the icons on the address bar located at the _top_ of the screen evade my attention. It is natural for me to check the info on the HUD (a la cVim) than to read the badge.

**Comment by saheel1115** [comment]
Oh, by the way, if you guys go the badge way, can you please consider changing the letter for Find mode to "F" or something. The "/" and "I" (of Insert mode) are just too confusing, apart from the general readability problem of them being small in size.

**Comment by saheel1115** [comment]
How about implementing the HUD and giving the user the option to hide/disable it? And you can keep the badges too for those who like icons. Yes, I understand that might take up more memory.

If you guys go the badge way, can you please consider changing the letter for Find mode to "F" or something. The "/" and "I" (of Insert mode) are just too confusing, apart from the general readability problem of them being small in size (should I start a new issue on this?). Also, the "/" doesn't make sense if I remap the Find mode key, so even for this reason it is perhaps better to keep a "F" for Find mode rather than a "/".

Not sure how much a user's views matter on such things, but here's 2 cents anyway:

1. Screen real estate is really precious for me on my 13 inch screen. Thus, many a times I prefer to use Full Screen, which hides the icon and leaves me with no information on Vimium mode. Also, I disagree with the point of real estate being made against HUD. I feel a small occasional HUD pop-up does not take up so much real estate. This is very subjective, though. Nonetheless, from an utilitarian POV, space freed by Full Screen - space taken by HUD >> 0.

2. I prefer to remove badges and keep my address bar clean, otherwise there'll be a plethora of them. Currently, I have badges for AdBlock and Vimium because I don't have a choice. If you guys decide to go the HUD way, I'll surely hide the Vimium icon.

3. This is subjective but reading the badge is definitely tougher for me, colored or otherwise. Being an Emacs user (and this must be true for Vim users as well), I am just so used to reading the status bar that the icons on the address bar located at the _top_ of the screen evade my attention. It is natural for me to check the info on the HUD (a la cVim) than to read the badge.

**Comment by leoj3n** [comment]
The badge is hard to read. I would prefer having the option to hide the Vimium badge. Perhaps Vimium could be smart enough to know it is hidden and switch to HUD mode in this case.

**Comment by mrmr1993** [comment]
> Perhaps Vimium could be smart enough to know it is hidden

Chrome deliberately stops us from doing this, so we can't pester people to show the browser action when they've chosen to hide it. Which, ironically, some of us seem to want to do anyway.

> switch to HUD mode in this case

If we default to HUD, then all of these problems go away. There would still be the question of whether to have

* badges always on
  - I think this will annoy some users who haven't disabled the icon already.
* badges only in absense of the HUD
  - we could let the user completely disable the HUD in the options (and mention that we'll use badges instead in the side comment). I think this hits the sweet spot.
* no badges

**Comment by mrmr1993** [comment]
A decision was made on this discussion was ended by #1570; it can probably be closed now.
